hotosm/fAIr
# Feature/vector tiles training files

**Blocked by #289**

## What does this PR do ? 

- Adds base model to models table , Default is RAMP but choices are Yolo and Ramp 


New Model table : 

![image](https://github.com/user-attachments/assets/e93d5d3f-13b0-4101-9720-8cfb0541a98c)



- Adds vector tile generation logic for the aois and labels in each training checkpoint 

![image](https://github.com/user-attachments/assets/e2480469-1128-415d-88df-3b3fde284c13)


This is how pmtiles will look like with AOI and labels + Queryable attribute 

- Adds banner table with its CURD API 

Table : 
![image](https://github.com/user-attachments/assets/d2259917-862f-4205-93d6-618456df04f1)

CURD: 

![image](https://github.com/user-attachments/assets/98f16ead-923d-49ee-b9e8-3b74abfa6125)


- Implements basic authentication of is user staff ? is user admin ? is this object belongs to user ? and is user logged in 

- Adds training centroid option when training is finished 
- Refactors all user related information from '`created_by`' or '`approved_by`' to user --- To maintain consistency 
- Adds searching  by id feature to models endpoint 

eg : https://fair-dev.hotosm.org/api/v1/model/?search=12 

- Adds feedback count to training object get endpoint directly 

eg : https://fair-dev.hotosm.org/api/v1/training/20/ 

- Improves admin view to change permission of users for admin -: Adds search , sort and update permission 
- Adds the kpi stats endpoint 

eg: https://fair-dev.hotosm.org/api/v1/kpi/stats/ 

## Consideration 

This should be the stable version , centroid which is being stored after each training is done so that it can be fetched when training is published and linked to model . This PR includes heavy refactor , please be cautious ! 

Deletion is not allowed for anyone until and unless it is not your object , only staffs/ admin are allowed to delete 
POST request is allowed if user is not associated to particular request : for eg - you can still create new training 

Django authentication is extended and followed the same standard.
Banner message is a textfield which can store raw .md 

## How to test ? 


- Test Pmtiles : https://fair-dev.hotosm.org/api/v1/workspace/download/dataset_286%2Foutput%2Ftraining_638%2Fmeta.pmtiles/ , Provide your OSM authentication 

- Try deleting some objects which is not yours 
- Try publishing the training of the model which is not yours , you should still be able to submit the training but not publish it ! 
- Make sure , this model api is searchable with model id : https://fair-dev.hotosm.org/api/v1/model/?search=12  
- Make sure total feedback count is being returned from this api https://fair-dev.hotosm.org/api/v1/training/20/ 
- Test KPI stats https://fair-dev.hotosm.org/api/v1/kpi/stats/ 
- Try publishing, disabling or deleting  banner and making sure it is reflected on the API https://fair-dev.hotosm.org/api/v1/banner/?limit=1 
- Try creating banner via API , you should not be allowed to the system until and unless you are a staff / an admin 


## TODO : 

- Publish .pmtiles to s3 and add cache in s3 bucket requests to render pmtiles faster
